[OpenMP] Fix preprocessor mismatches between include and usages of hwloc

pearzt opened the pull request:
Fix https://github.com/llvm/llvm-project/issues/156679

There is a mismatch between the preprocessor guards around the include of `hwloc.h` and those protecting its usages, leading to build failures on Darwin: https://github.com/spack/spack-packages/pull/1212

The suggested change introduces ~`KMP_INCLUDES_HWLOC`~`KMP_HWLOC_ENABLED` that reflects whether hwloc is actually used. ~I realize that having both `KMP_USE_HWLOC` and `KMP_HWLOC_ENABLED` could be confusing, so I'm open to alternative naming suggestions.~

pearzt committed:
[OpenMP] Fix preprocessor mismatches between include and usages of hwloc

mjklemm reviewed (changes requested):
I think `KMP_INCLUDES_HWLOC` does not correctly describe what it actually means. `KMP_USE_HWLOC` is to enable hwloc support in general, bearing in mind that additional conditions may disable it again.  I think `KMP_HWLOC_ENABLED` might be a better choice, because it then describes that a) `hwloc.h` has been included, and b) that the hwloc-based affinity code is also enabled.

pearzt committed:
[OpenMP] Use KMP_HWLOC_ENABLED and fix #endif comments

KMP_HWLOC_ENABLED expresses whether hwloc is included *and*
hwloc-based affinity is enabled.

pearzt commented:
@mjklemm I've now changed it back to use `KMP_HWLOC_ENABLED`. As long as nobody else has a better naming suggestion or other objections, I think this is ready to move forward, right?

mjklemm reviewed (approved):
LGTM. Thanks for making the changes!
